### PR TITLE
Fix for automatic reporting of crashes

### PR DESCRIPTION
**Pull request by peterBrxwn**
Without this fix, some crash reports are not automatically sent to server when you `enableCrashReporting` from config.

### CODE REVIEW

**Comment by peterBrxwn** [comment]
[[Flutter] Fix automatic crash filtering bug](https://trello.com/c/jBdbAV4D/1916-flutter-fix-automatic-crash-filtering-bug)

**Review by AliRKat — APPROVED** [review]

**Review by turtledreams — CHANGES_REQUESTED** [review]
I combined your tests to a single file because it was needless for them to be in separate files. you have to write the bottom validation part though, i did not touch that.
What we really want to test is unhandled crashes. how can we achieve that?

**Comment by peterBrxwn** [comment]
I have fixed the tests.

File: example/lib/main.dart [hunk 1]
```diff
@@ -22,15 +22,13 @@ void main() {
   /// This or a similar call needs to added to catch and report Dart Errors to Countly,
```
Comment (turtledreams) [inline example/lib/main.dart:22]: these comments should also change

File: lib/src/countly_flutter.dart [hunk 2]
```diff
@@ -257,6 +257,11 @@ class Countly {
       }
     }
     if (config.enableUnhandledCrashReporting != null) {
+      FlutterError.onError = _recordFlutterError;
```
Comment (turtledreams) [inline lib/src/countly_flutter.dart:260]: please add comments here describing what they do for future developers

File: example/lib/page_crash_reporting.dart [hunk 3]
```diff
@@ -73,6 +67,13 @@ class CrashReportingPage extends StatelessWidget {
             MyButton(text: 'Throw Native Exception', color: 'orange', onPressed: throwNativeException),
             MyButton(text: 'Record Exception Manually', color: 'teal', onPressed: recordExceptionManually),
             MyButton(text: 'Divided By Zero Exception', color: 'teal', onPressed: dividedByZero),
+            MyButton(
```
Comment (turtledreams) [inline example/lib/page_crash_reporting.dart:70]: no new buttons

File: example/lib/page_crash_reporting.dart [hunk 4]
```diff
@@ -43,16 +43,10 @@ class CrashReportingPage extends StatelessWidget {
...
-      int firstInput = 20;
-      int secondInput = 0;
-      int result = firstInput ~/ secondInput;
-      print('The result of $firstInput divided by $secondInput is $result');
-    } catch (e, s) {
-      print('Exception occurs: $e');
-      print('STACK TRACE\n: $s');
-      Countly.logExceptionEx(e as Exception, true, stacktrace: s);
-    }
+    int firstInput = 20;
```
Comment (turtledreams) [inline example/lib/page_crash_reporting.dart:46]: still not reversed

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,6 @@
+## X.X.X
```
Comment (turtledreams) [inline CHANGELOG.md:1]: changelog changes

**Review by turtledreams — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 5 above)
Comment (peterBrxwn) [reply CHANGELOG.md:1]: added

File: example/lib/page_crash_reporting.dart (hunk 4 above)
Comment (peterBrxwn) [reply example/lib/page_crash_reporting.dart]: reversed

File: lib/src/countly_flutter.dart (hunk 2 above)
Comment (peterBrxwn) [reply lib/src/countly_flutter.dart:260]: okay

File: example/lib/main.dart (hunk 1 above)
Comment (peterBrxwn) [reply example/lib/main.dart]: DOne

File: example/lib/page_crash_reporting.dart (hunk 3 above)
Comment (peterBrxwn) [reply example/lib/page_crash_reporting.dart:74]: This new button is to specifically test that async errors is caught by `PlatformDispatcher.instance.onError`. Formally, it was caught by runZoned, but now we've removed that

Comment (peterBrxwn) [reply example/lib/page_crash_reporting.dart:74]: Should I remove it?

Comment (turtledreams) [reply example/lib/page_crash_reporting.dart:74]: no I mean there are no new buttons to test these errors without a try catch block

File: README.md [hunk 6]
```diff
@@ -48,10 +48,8 @@ dependencies:
 import 'package:countly_flutter/countly_flutter.dart';
 
 void main() {
-  // If you want to catch Dart errors, run your app inside a Zone and pass Countly.recordDartError as the onError parameter to it.
-  runZonedGuarded<void>(() {
-    runApp(MaterialApp(home: const MyApp()));
-  }, Countly.recordDartError);
+  // If you want to catch Dart errors, call enableCrashReporting before you initialize your Countly SDK.
```
Comment (turtledreams) [inline README.md:51]: previously catching dart errors needed extra effort and it was noted. Is it still the same? If not why we mention Dart errors here.?

File: example/lib/page_crash_reporting.dart (hunk 3 above)
Comment (turtledreams) [reply example/lib/page_crash_reporting.dart:74]: I told you to add a button for each unique crash

**Review by turtledreams — CHANGES_REQUESTED** [review]

**Review by turtledreams — CHANGES_REQUESTED** [review]
changelog has to be fixed. way too much implementation details are given. just write what the user should know, like:
added support for X type of exception capturing
mitigated an issue where automatic crash reporting did not work
etc

Remove Dart error comments I mentioned. that is not relevant anymore

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,6 @@
 ## X.X.X
-* Mitigated an issue where some crash reports are not automatically sent when you `enableCrashReporting` from config.
-* Removed all reference to `runZonedGuarded`. Dart and async errors are caught using `PlatformDispatcher.instance.onError` if crashReporting is enabled.
+* Resolved an issue where certain crash reports were not automatically sent.
+* Removed all reference to `runZonedGuarded`.
```
Comment (turtledreams) [inline CHANGELOG.md:3]: yea this was the detail I had in mind :D they dont have to know we removed mentions of this. 

But you fixed 2 things:
1 automatic crash reporting did not include flutter framework level errors with the new config option we introduced but u fixed it
2. we were not capturing async dart errors but now we do

**Review by turtledreams — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (turtledreams) [inline README.md:51]: this is the thing you should remove

File: scripts/no-push-files/README.md [hunk 8]
```diff
@@ -47,10 +47,8 @@ dependencies:
 import 'package:countly_flutter/countly_flutter.dart';
 
 void main() {
-  // If you want to catch Dart errors, run your app inside a Zone and pass Countly.recordDartError as the onError parameter to it.
-  runZonedGuarded<void>(() {
-    runApp(MaterialApp(home: const MyApp()));
-  }, Countly.recordDartError);
+  // If you want to catch Dart errors, call enableCrashReporting before you initialize your Countly SDK.
```
Comment (turtledreams) [inline scripts/no-push-files/README.md:50]: also remove this comment. if they want to capture any error the method is same. no need to single out dart

**Review by turtledreams — CHANGES_REQUESTED** [review]

File: scripts/no-push-files/README.md (hunk 8 above)
Comment (peterBrxwn) [reply scripts/no-push-files/README.md]: Done

File: README.md (hunk 6 above)
Comment (peterBrxwn) [reply README.md]: Done

File: CHANGELOG.md (hunk 7 above)
Comment (peterBrxwn) [reply CHANGELOG.md]: Done

**Review by turtledreams — APPROVED** [review]

**Pull request merged**